### PR TITLE
Cross-Validation Tools

### PR DESCRIPTION
**Pull request by ahwillia**
This is my first pass at a set of cross-validation tools for twpca. The core code is in `twpca/crossval.py` and some rough examples are in `examples/crossval-demo/` (check out the notebooks in that folder).

Probably not ready to merge, still looking for feedback on the cross validation API and organization, as well as on the scope of the tools we want to offer is. For example, do we want to include plotting functions? (At the moment all the results are packaged into rather large and clunky dicts, and visualization would be helpful.)

### CODE REVIEW

**Comment by ahwillia** [comment]
@poolio - this is ready for review. See `crossval-demo.ipynb` for usage.

image

image

image

File: twpca/crossval.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+import tensorflow as tf
+import itertools
+from tqdm import tqdm
+from .model import TWPCA
+from .regularizers import curvature
+from .utils import stable_rank
+
+__all__ = ['cross_validate', 'hyperparam_search', 'err_per_component']
+
+def k_fold(N, K):
```
Comment (poolio) [inline twpca/crossval.py]: could we give these names like k_fold_iter or k_fold_iterator

File: twpca/crossval.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    """K-fold cross validation iterator
+
+    Args:
+        N: int, number of datapoitns
+        K: int, number of folds
+    """
+    idx = np.arange(N)
```
Comment (poolio) [inline twpca/crossval.py]: I would advocate for using additional randomness here, not just striding. An easy modification would be:
`idx = np.random.permutation(N)`

File: twpca/crossval.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        yield train_idx, test_idx
+
+def leave_n_out(N, K):
+    """Leave K out cross validation iterator
+
+    Args:
+        N: int, number of datapoitns
+        K: int, number to leave out in training set
+    """
+    idx = np.arange(N)
```
Comment (poolio) [inline twpca/crossval.py]: `np.random.permutation` ?

File: twpca/crossval.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        N: int, number of datapoitns
+        K: int, number of folds
+    """
+    idx = np.arange(N)
+    for k in range(K):
+        train_idx = idx[idx % K != k]
+        test_idx = idx[idx % K == k]
+        yield train_idx, test_idx
+
+def leave_n_out(N, K):
```
Comment (poolio) [inline twpca/crossval.py]: change name to `leave_k_out` or change docstring and variable names below

File: twpca/crossval.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        yield train_idx, test_idx    
+
+def cross_validate(model, data, method, K, max_crossval=np.inf, **fit_kw):
+    """Runs specified cross-validation method.
+
+    Args:
+        model: TWPCA model instance
+        data: array-like (n_trials x n_timepoints x n_neurons)
+        method: string specifying method {'kfold', 'leaveout'}
+        K: int, crossvalidation parameter
```
Comment (poolio) [inline twpca/crossval.py:50]: missing max_crossval, fit_kw docstring

File: twpca/crossval.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    elif method == 'leaveout':
+        partitions = leave_k_out(data.shape[2], K)
+    else:
+        raise ValueError('Cross-validation method not recognized.')
+
+    nfits = 0
+    results = []
+    for train, test in partitions:
+        # partition dataset
+        traindata = np.atleast_3d(data[:, :, train])
```
Comment (poolio) [inline twpca/crossval.py]: do you need the atleast_3d? it looks like train/test will always be lists

File: twpca/crossval.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        testdata = np.atleast_3d(data[:, :, test])
+
+        # fit model to training set
+        tf.reset_default_graph()
```
Comment (poolio) [inline twpca/crossval.py]: can you add a TODO(ben): better session management

File: twpca/crossval.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        sess = tf.Session()
+        model.fit(traindata, sess=sess, **fit_kw)
+        # TODO - warmstart fits?
```
Comment (poolio) [inline twpca/crossval.py]: While this would speed things up, I think it would reduce the variability across runs, so might not be worth doing here.

File: twpca/crossval.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+        # compile results
+        results.append(
+            {'params': model.params,
+             'warped_rank': np.array(warped_rank),
+             'unwarped_rank': np.array(unwarped_rank),
+             'test_idx': test,
+             'train_idx': train,
+             'train_error': model._sess.run(model.recon_cost),
+             'test_error': np.sum(resid) / len(testdata.ravel()),
```
Comment (poolio) [inline twpca/crossval.py]: have you checked that train/test error are computed the same?

File: twpca/crossval.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        # assess dimensionality of warped vs unwarped testdata
+        warped_testdata = model.transform(testdata)
+        warped_rank, unwarped_rank = [], []
+        for n in range(testdata.shape[2]):
+            warped_rank.append(stable_rank(warped_testdata[:, :, n]))
+            unwarped_rank.append(stable_rank(testdata[:, :, n]))
+
+        # evaluate test error
+        time_factors = sess.run(model._warped_time_factors).reshape(-1, model.n_components)
+        _, resid, _, _ = np.linalg.lstsq(time_factors, testdata.reshape(-1, len(test)))
```
Comment (poolio) [inline twpca/crossval.py]: this should really be a function in the model, either transform() or apply() or something

File: twpca/crossval.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+            break
+
+    return results
+
+def hyperparam_search(data, n_components, warp_scales, time_scales,
+                      warp_reg=None, time_reg=None,
+                      crossval_method='kfold', K=5, max_crossval=np.inf,
+                      fit_kw=dict(lr=(1e-1, 1e-2), niter=(250, 500), progressbar=False),
+                      **model_kw):
+    """Performs cross-validation over number of components, warp regularization scale, and
```
Comment (poolio) [inline twpca/crossval.py]: add docstring here. what's expected format of fit_kw and model_kw? List of parameters that you're going to take the cross product over?

File: twpca/crossval.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        'warp_scale': [],
+        'time_scale': [],
+        'crossval_data': [],
+        'mean_test': [],
+        'mean_train': [],
+        'mean_dim_change': []
+    }
+
+    # run cross-validation for all specified hyperparameters
+    for nc, ws, ts in zip(tqdm(n_components), warp_scales, time_scales):
```
Comment (poolio) [inline twpca/crossval.py:182]: does this work? I think tqdm() expects an iterator and n_components is an integer

File: twpca/crossval.py (hunk 11 above)
Comment (poolio) [reply twpca/crossval.py]: also it looks like warp_reg and time_reg have a weird expected format of functions that take a penalty amount as an argument, please doc!

File: twpca/crossval.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        results['warp_scale'].append(ws)
+        results['time_scale'].append(ts)
+        results['crossval_data'].append(_result)
+        results['mean_test'].append(np.mean([r['test_error'] for r in _result]))
+        results['mean_train'].append(np.mean([r['train_error'] for r in _result]))
+        results['mean_dim_change'].append(np.mean([r['warped_rank']-r['unwarped_rank'] for r in _result]))
+
+    return results
+
+def err_per_component(data, component_range,
```
Comment (poolio) [inline twpca/crossval.py]: error_per_component?

File: twpca/datasets.py [hunk 14]
```diff
@@ -54,3 +54,63 @@ def jittered_neuron(t=None, feature=None, n_trial=61, jitter=1.0, gain=0.0, nois
...
+        trial_event_times = np.cumsum(event_gap + jitter)
+        events[trial_idx, trial_event_times] = 1.0
+    avg_event = np.zeros(n_time)
+    avg_event[np.cumsum([event_gap] * n_events)] = 1.0
+
+    # Convolve latent events with an exponential filter
+    impulse_response = np.exp(-np.arange(n_time)/float(tau))
+    impulse_response /= impulse_response.sum()
+
+    latents = np.array([np.convolve(e, impulse_response, mode='full')[:n_time] for e in events])
```
Comment (poolio) [inline twpca/datasets.py:103]: can use scipy.ndimage.filters.convolve1d to do this all at once instead of in a loop

File: twpca/datasets.py [hunk 15]
```diff
@@ -54,3 +54,63 @@ def jittered_neuron(t=None, feature=None, n_trial=61, jitter=1.0, gain=0.0, nois
...
+
+    # Coupling from one dimensional latent state to each neuron
+    readout_weights = np.random.rand(n_neuron) + 0.1
```
Comment (poolio) [inline twpca/datasets.py]: where's this 0.1 come from?

File: twpca/model.py [hunk 16]
```diff
@@ -59,13 +59,14 @@ def __init__(self, n_components, shared_length=None,
...
 
-    def fit(self, X, optimizer=tf.train.AdamOptimizer, niter=1000, lr=1e-3, sess=None):
+    def fit(self, X, optimizer=tf.train.AdamOptimizer, niter=1000, lr=1e-3, sess=None, progressbar=True):
         """Fit the twPCA model
 
         Args:
             X: 3D numpy array with dimensions [n_trials, n_timepoints, n_neurons]
             optimizer (optional): a tf.train.Optimizer class (default: AdamOptimizer)
             niter (optional): number of iterations to run the optimizer for (default: 1000)
+            progressbar (optional): whether to print a progressbar (default: True)
```
Comment (poolio) [inline twpca/model.py]: fix order, sess first, progressbar second

**Review by poolio — COMMENTED** [review]
High level: great addition! It would be nice if we had options for cross-validation along any axis (time, neuron, trial) instead of just neurons.

Didn't get a chance to look at notebooks yet, will do a second pass for that tomorrow.

File: twpca/crossval.py (hunk 4 above)
Comment (ahwillia) [reply twpca/crossval.py]: 👍

File: twpca/crossval.py (hunk 3 above)
Comment (ahwillia) [reply twpca/crossval.py]: I was back and forth about which would be better. If we're doing this over multiple hyperparams we might want the same random permutation... I think I'll fix by passing a seed.

File: twpca/crossval.py (hunk 1 above)
Comment (ahwillia) [reply twpca/crossval.py]: done.

File: twpca/crossval.py (hunk 2 above)
Comment (ahwillia) [reply twpca/crossval.py]: done.

File: twpca/crossval.py (hunk 3 above)
Comment (ahwillia) [reply twpca/crossval.py]: done.

File: twpca/crossval.py (hunk 5 above)
Comment (ahwillia) [reply twpca/crossval.py:50]: done.

File: twpca/crossval.py (hunk 6 above)
Comment (ahwillia) [reply twpca/crossval.py]: Nope, good catch.

File: twpca/crossval.py (hunk 7 above)
Comment (ahwillia) [reply twpca/crossval.py]: done.

File: twpca/crossval.py (hunk 8 above)
Comment (ahwillia) [reply twpca/crossval.py]: deleted.

File: twpca/crossval.py (hunk 12 above)
Comment (ahwillia) [reply twpca/crossval.py:182]: strangely enough it does seem to work. But replaced with `trange`

File: twpca/datasets.py (hunk 14 above)
Comment (ahwillia) [reply twpca/datasets.py:103]: done.

File: twpca/datasets.py (hunk 15 above)
Comment (ahwillia) [reply twpca/datasets.py]: deleted.

File: twpca/model.py (hunk 16 above)
Comment (ahwillia) [reply twpca/model.py]: done.

File: twpca/crossval.py (hunk 10 above)
Comment (ahwillia) [reply twpca/crossval.py]: `transform()` applies the warps. Do you want me to make `model.apply()`?

File: twpca/crossval.py (hunk 9 above)
Comment (ahwillia) [reply twpca/crossval.py]: Yes, but perhaps better to put this inside the model class.

File: twpca/crossval.py (hunk 12 above)
Comment (ahwillia) [reply twpca/crossval.py:182]: Ah my bad - `n_components` is actually passed in as a list. See the updated docstring.

**Comment by ahwillia** [comment]
@poolio - ready to review again.

File: twpca/datasets.py (hunk 14 above)
Comment (poolio) [reply twpca/datasets.py:103]: it looks like you imported but still aren't using convolve1d

File: twpca/model.py [hunk 17]
```diff
@@ -196,6 +197,40 @@ def transform(self, X=None):
 
         return self._sess.run(warp.warp(X_tf, self._inv_warp))
 
+    def predict(self, X=None):
+        """Return model prediction of activity on each trial.
+
```
Comment (poolio) [inline twpca/model.py:202]: can you add docstring for the expected size of X and a check that the size of X matches the temporal factors

File: twpca/crossval.py [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+        results['mean_test'].append(np.mean([r['test_error'] for r in _result]))
+        results['mean_train'].append(np.mean([r['train_error'] for r in _result]))
+        results['mean_dim_change'].append(np.mean([r['warped_rank']-r['unwarped_rank'] for r in _result]))
+
+    return results
+
+def error_per_component(data, component_range,
+                        fit_kw=dict(lr=(1e-1, 1e-2), niter=(250, 500), progressbar=False),
+                        **model_args):
+    """Compares vanilla PCA to twPCA searching over number of components
```
Comment (poolio) [inline twpca/crossval.py]: add docstring for args/returns

File: twpca/crossval.py [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+    return results
+
+def hyperparam_search(data, n_components, warp_scales, time_scales,
+                      warp_reg=None, time_reg=None,
+                      crossval_method='kfold', K=5, max_fits=np.inf,
+                      fit_kw=dict(lr=(1e-1, 1e-2), niter=(250, 500), progressbar=False),
+                      **model_kw):
+    """Performs cross-validation over number of components, warp regularization scale, and
+    temporal regularization scale.
+
```
Comment (poolio) [inline twpca/crossval.py:125]: can you document somewhere how the keywoard arguments are combined? typically these kind of functions take the outer product over all parameters, but here you're taking them both in as lists of parameter settings to try

**Review by poolio — CHANGES_REQUESTED** [review]
a few more minor changes to the code. almost there :)

File: twpca/crossval.py (hunk 18 above)
Comment (ahwillia) [reply twpca/crossval.py]: This isn't exported. I'm still playing with the functionality I want - will remove for now.

File: twpca/model.py (hunk 17 above)
Comment (ahwillia) [reply twpca/model.py:202]: done

File: twpca/crossval.py (hunk 19 above)
Comment (ahwillia) [reply twpca/crossval.py:125]: Sort of verbose but the docstring has an example use case
